### ISSUE TITLE
Failed distributed query with complex arrayJoin() expression

### ISSUE THREAD

**Issue by alex-zaitsev** [yandex/ClickHouse#2868]
Easiest way to reproduce:
```
select arrayJoin(arrayFilter(x->length(x) in (select number from system.numbers limit 5), attribute.names)) from remote('host1,host2', system.dictionaries);

SELECT arrayJoin(arrayFilter(x -> (length(x) IN 
    (
        SELECT number
        FROM system.numbers 
        LIMIT 5
    )), attribute.names))
FROM remote('host1,host2', system.dictionaries) 

Received exception from server (version 18.6.54401):
Code: 10. DB::Exception: Received from chdw1-1.sqc04:9000, 10.253.59.9. DB::Exception: Not found column arrayJoin(arrayFilter(lambda(tuple(x), in(length(x), _subquery1)), attribute.names)) in block. There are only columns: name, origin, type, key, attribute.types, bytes_allocated, query_count, hit_rate, element_count, load_factor, creation_time, source, last_exception, arrayJoin(arrayFilter(lambda(tuple(x), in(length(x), _subquery1__subquery_16466317254568153445_12029335626405941901)), attribute.names)). 
```
Note, it can not match essentially identical columns:
arrayJoin(arrayFilter(lambda(tuple(x), in(length(x), _subquery1)), attribute.names))
arrayJoin(arrayFilter(lambda(tuple(x), in(length(x), _subquery1__subquery_16466317254568153445_12029335626405941901)), attribute.names))

Error is reproducible on any array in the distributed query.
Error is reproducible in 1.1.54394 and 18.1. I am not sure about 18.6 and above -- we do not have a homogeneous cluster to test.
There is no error in 1.1.54385.

**Comment by SaltTan** [comment]
The issue is present in 18.10.3
```
:) SELECT arrayFilter(x -> x IN (select 42), [42]) from remote('localhost,chdw1-1.sde317', system.one)

SELECT arrayFilter(x -> (x IN
    (
        SELECT 42
    )), [42])
FROM remote('localhost,some_other_host', system.one)

┌─arrayFilter(lambda(tuple(x), in(x, _subquery1)), [42])─┐
│ [42]                                                   │
└────────────────────────────────────────────────────────┘
Received exception from server (version 18.10.3):
Code: 10. DB::Exception: Received from localhost:9000, ::1. DB::Exception: Not found column arrayFilter(lambda(tuple(x), in(x, _subquery1)), [42]) in block. There are only columns: dummy, arrayFilter(lambda(tuple(x), in(x, _subquery1__subquery_3113076859249138518_4974159629285336334)), [42]).

1 rows in set. Elapsed: 0.050 sec.
```

**Comment by vmanyushin** [comment]
```
SELECT
    user_id,
    if(user_id NOT IN
    (
        SELECT user_id
        FROM events.finance
        WHERE project_id = 282
        GROUP BY user_id
    ), 'yes', 'no') AS project282
FROM events.login
LIMIT 100

2018.08.21 15:53:49.196117 [ 79 ] <Error> executeQuery: Code: 10, e.displayText() = DB::Exception: Not found column if(notIn(user_id, _subquery1), 'yes', 'no') in block. There are only columns: user_id, if(notIn(user_id, _subquery1__subquery_2798086608428224864_7530129903691534915), 'yes', 'no'), e.what() = DB::Exception (from 127.0.0.1:38180) (in query: SELECT user_id, if( user_id not in (SELECT user_id FROM events.finance WHERE project_id = 282 GROUP BY user_id), 'yes', 'no') as project282 FROM events.login LIMIT 100), Stack trace:

0. clickhouse-server(StackTrace::StackTrace()+0x16) [0x4b84796]
1. clickhouse-server(DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)+0x1f) [0x266aeef]
2. clickhouse-server(DB::Block::getByName(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0x17f) [0x3cd601f]
3. clickhouse-server(DB::RemoteBlockInputStream::readImpl()+0x3ee) [0x3cf867e]
4. clickhouse-server(DB::IProfilingBlockInputStream::read()+0x20f) [0x3ce8bff]
5. clickhouse-server(DB::LimitBlockInputStream::readImpl()+0x36a) [0x3ce9c0a]
6. clickhouse-server(DB::IProfilingBlockInputStream::read()+0x20f) [0x3ce8bff]
7. clickhouse-server(DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0>::Handler, (DB::StreamUnionMode)0>::loop(unsigned long)+0xf3) [0x43cf9a3]
8. clickhouse-server(DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0>::Handler, (DB::StreamUnionMode)0>::thread(MemoryTracker*, unsigned long)+0x1f0) [0x43d0250]
9. clickhouse-server() [0x5290a7f]
10. /lib64/libpthread.so.0(+0x7e25) [0x7f5898ec6e25]
11. /lib64/libc.so.6(clone+0x6d) [0x7f58984cdbad]

```

is this the same error? version 18.10.6

**Comment by alex-zaitsev** [comment]
Looks fixed in 18.12.13

https://github.com/yandex/ClickHouse/pull/3087
